content_handler/application_impl.cc

```diff
@@ -6,25 +6,46 @@
 
 #include <utility>
 
+#include "apps/dart_content_handler/zip/unzipper.h"
+#include "flutter/content_handler/runtime_holder.h"
 #include "lib/ftl/logging.h"
 #include "mojo/public/cpp/application/connect.h"
 
 namespace flutter_content_handler {
+namespace {
+
+constexpr char kSnapshotKey[] = "snapshot_blob.bin";
+
+std::vector<char> ExtractSnapshot(std::vector<char> bundle) {
+  zip::Unzipper unzipper(std::move(bundle));
+  return unzipper.Extract(kSnapshotKey);
+}
+
+}  // namespace
 
 ApplicationImpl::ApplicationImpl(
     mojo::InterfaceRequest<mojo::Application> application,
     mojo::URLResponsePtr response)
-    : binding_(this, std::move(application)),
-      initial_response_(std::move(response)) {}
+    : binding_(this, std::move(application)) {
+  drainer_.reset(new glue::DrainDataPipeJob(
```

Comment:
Yeah, we're inconsistent about which pattern we used.  I'm coming around to thinking this pattern is better.  The benefit is that if `ApplicationImpl` dies during the drain, we'll destroy the `drainer_`, which will cancel the job.  If we used the "callback causes delete" pattern, then the drain would continue until it got all the data, which it would proceed to drop on the floor and then delete itself.

We probably should move more things to this pattern of the client holding an owning reference to the job.  The client can always convert that to "callback causes delete", but once you have a self-owned job, there's no way to convert it to the other pattern because there's no pathway for canceling the work.